### DISCUSSION TITLE
How to SSH to a device from internet.

### DISCUSSION THREAD

**Discussion by ankur112358** [code-name-57/iot-drone#2] [General]
How to SSH to a device from internet.

Ref:
[solution1](https://www.quora.com/How-do-I-connect-through-SSH-to-a-device-using-Verizon-4G-LTE-from-the-Internet)
[answer - reverse ssh](http://www.thirdway.ch/En/projects/raspberry_pi_3g/index.php)
[descriptive solution vpn and tunneling](https://android.stackexchange.com/questions/203996/how-to-connect-to-android-through-ssh-over-3g-4g-public-ip)

**Comment by ankur112358** [comment]
How web app on aws get connected to domain name without static ip.

**Comment by ankur112358** [comment]
Now the question is how does VPN solution scales to multiple drones and users?

**Comment by ankur112358** [comment]
Hamachi can be used - it is a VPN software, to create LAN over WAN(the internet)

**Reply by Macbull** [reply]
Found a relevant answer on https://forums.whirlpool.net.au/archive/9nky4y43
Uses VPN and port forwarding. 
> Most 4G connections share the same IP between hundreds of users so you can't use DDNS to accept inbound traffic.
> 
> The way I have done it is using Mikrotik
> 
> (1) Use a Mikrotik router at home/work.
> (2) Use a Mikrotik router on your 4G connection.
> (3) Open a SSTP VPN from the 4G to home router
> (3) Set up appropriate NAT/port forwarding settings on both sides.
> (4) Your home or work's router will then forward packets onto the 4G device.
> 
> That way you don't expose your 4G modem to the public internet with a public IP.
> 
> Alternatively you can install Mikrotik CHR on the cheapest possible VPS server and do the same thing (so you don't have to change your home or work router.
> 
> You would need to Google a lot to get it going, you can even buy it from Duxtel who have Australian support if you get stuck.
> 
> This way you own and control the link without using a third party VPN.

**Reply by Macbull** [reply]
DynDNS service provides a way around dynamic IP but still relies on public ip of the actual drone and doesn't work with NAT by ISP of 4g connections. Can work pretty well for 4g(NAT)-Broadband Connection(Dynamic IP).

Router on broadband side will need to have port forwarding, and connection can be established using reverse shell. Where ground station will be listening for connection and can be reached through dynDNS address. 

The dyndns address can be stored in database associated with paired user account. 
This might also be more secure since we are not storing public IP of the drone.

**Reply by Macbull** [reply]
In case of 4g-4g connection, VPN service seems to be the best option in terms of latency. SSH tunneling is also an option but slower. 
There is also an option to use a server as Man-in-Middle which will receive telemetry message from drone, forward them to ground station and vice versa. This will be slow, since forwarding happens in application layer. But this can be used to log telemetry message on cloud for analysis, sharing, team operation or share it with FAA (Refer Remote ID regulation in US).
This will also enable easy development of web based ground station.

**Reply by Macbull** [reply]
DNS Server explained 
https://www.cloudflare.com/learning/dns/what-is-a-dns-server/
Various types of VPN explained  
https://www.cactusvpn.com/beginners-guide-to-vpn/what-is-openvpn/#:~:text=OpenVPN%20is%20both%20an%20open,it%20works%20across%20multiple%20platforms.
Reverse Shell explained
https://www.acunetix.com/blog/web-security-zone/what-is-reverse-shell/

**Reply by Macbull** [reply]
CDN might be possible solution.
https://aws.amazon.com/cloudfront/

DNS service by amazon 
https://aws.amazon.com/route53/

Virtual private cloud gives complete control over infrastructure of virtual private network.
https://aws.amazon.com/vpc/?vpc-blogs.sort-by=item.additionalFields.createdDate&vpc-blogs.sort-order=desc

**Comment by ankur112358** [comment]
[remote it](https://www.youtube.com/watch?v=_B8E1dE5kW4) is also a viable solution. And it looks pretty easy with an interface for multiple devices.

With this approach:-
* performing complete connection via code might be some work but it should not be a very hard challenge.
* performing complete connection via code might not be necessary. the ui/ux of remote it looks clean.
